### PR TITLE
Introducing.... LRController

### PR DESCRIPTION
**Pull request by vbvg2008**
LRController is unifying lrscheduler and reducelronplateau.

### CODE REVIEW

File: fastestimator/estimator/estimator.py [hunk 1]
```diff
@@ -143,13 +148,24 @@ def _sort_traces(self):
...
             state = {"mode": mode}
-            for epoch in signature_epochs:
+            for idx, epoch in enumerate(signature_epochs):
                 ds_iter = self.pipeline.dataset_schedule[mode].get_current_value(epoch)
+                if mode == "train":
+                    global_batch_size = self.pipeline.get_global_batch_size(epoch)
+                    if self.steps_per_epoch:
+                        max_steps = self.steps_per_epoch
+                    else:
+                        max_steps = min(self.pipeline.num_examples[mode]) // global_batch_size
```
Comment (TortoiseHam) [inline fastestimator/estimator/estimator.py]: do we do padding on the last batch in an epoch to fill it out? if so, don't we need to round up here?

Comment (vbvg2008) [reply fastestimator/estimator/estimator.py]: no padding

File: fastestimator/estimator/trace.py [hunk 2]
```diff
@@ -195,6 +195,95 @@ def on_epoch_end(self, state):
...
+
+
+class LRController(Trace):
+    def __init__(self,
+                 model,
+                 lr_schedule=None,
+                 reduce_on_eval=False,
+                 reduce_patience=10,
+                 reduce_factor=0.1,
+                 reduce_metric="loss",
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: why separate reduce_on_eval from reduce_metric? we can just infer reduce_on_eval from whether a non-None reduce_metric is provided

File: fastestimator/estimator/trace.py [hunk 3]
```diff
@@ -195,6 +195,95 @@ def on_epoch_end(self, state):
...
+        self.change_lr = False
+        self.wait = 0
+        if self.lr_schedule:
+            assert isinstance(self.lr_schedule, LRSchedule), "lr_schedule must be instance of LRSchedule"
+        if self.reduce_metric_mode == "min":
+            self.reduce_metric_best = np.Inf
+        elif self.reduce_metric_mode == "max":
+            self.reduce_metric_best = -np.Inf
+        else:
+            raise ValueError("reduce_metric_mode is either 'min' or 'max'")
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: "must be either 'min' or 'max' "

File: fastestimator/estimator/trace.py [hunk 4]
```diff
@@ -195,6 +195,95 @@ def on_epoch_end(self, state):
...
+            current_value = state[self.reduce_metric]
+            if self._is_better(current_value):
+                self.reduce_metric_best = current_value
+                self.wait = 0
+            else:
+                self.wait += 1
+                if self.wait >= self.reduce_patience:
+                    self.reduce_lr_ratio *= self.reduce_factor
+                    self.change_lr = True
+                    print("FastEstimator-LearningRateChanger: learning rate reduced by factor of {}".format(
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: "FastEstimator-LRController..."

File: fastestimator/estimator/trace.py [hunk 5]
```diff
@@ -792,7 +821,7 @@ def on_epoch_end(self, state):
                     for name, model in self.network.model.items():
                         model.set_weights(self.best_weights[name])
                 print("FastEstimator-EarlyStopping: '{}' triggered an early stop. Its best value was {} at epoch {}\
-                        ".format(self.monitored_key, self.best, state['epoch'] - self.wait))
+                        "                         .format(self.monitored_key, self.best, state['epoch'] - self.wait))
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: what's going on with this?

File: fastestimator/network/lrschedule.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
+import math
+
+import numpy as np
+
+
+class LRSchedule:
+    def __init__(self, schedule_mode):
+        self.schedule_mode = schedule_mode
+        self.total_epochs = None  #will be filled on runtime
```
Comment (TortoiseHam) [inline fastestimator/network/lrschedule.py]: "will be filled by runtime"

Comment (vbvg2008) [reply fastestimator/network/lrschedule.py]: fixed

File: fastestimator/estimator/trace.py (hunk 5 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: solar flare

File: fastestimator/estimator/trace.py (hunk 4 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: fixed

File: fastestimator/estimator/trace.py (hunk 3 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: fixed

File: fastestimator/estimator/trace.py (hunk 2 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: fixed.

File: apphub/image_classification/lenet_mnist.py [hunk 7]
```diff
@@ -42,7 +43,7 @@ def get_estimator(epochs=2, batch_size=32):
     # step 3.prepare estimator
     traces = [
         Accuracy(true_key="y", pred_key="y_pred", output_name='acc'),
-        ModelCheckpoint('./lenet_model', monitor_name='acc', save_best_only=True)
+        LRController(model_name="lenet", lr_schedule=CyclicLRSchedule())
```
Comment (TortoiseHam) [inline apphub/image_classification/lenet_mnist.py]: if you're removing ModelCheckpoint, need to clean up the import statement

Comment (vbvg2008) [reply apphub/image_classification/lenet_mnist.py]: not removing, i'm currently cleaning up the checkpoint code

File: fastestimator/estimator/trace.py [hunk 8]
```diff
@@ -195,6 +195,97 @@ def on_epoch_end(self, state):
...
+class LRController(Trace):
+    def __init__(self,
+                 model_name,
+                 lr_schedule=None,
+                 reduce_on_eval=False,
+                 reduce_patience=10,
+                 reduce_factor=0.1,
+                 reduce_mode="min",
+                 min_lr=1e-6):
+        if reduce_on_eval and isinstance(reduce_on_eval, str):
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: this can just be 
`if is instance(reduce_on_eval, str)`

Comment (TortoiseHam) [reply fastestimator/estimator/trace.py]: or even get rid of the outer if and do

super().__init__(inputs=reduce_on_eval if isinstance(reduce_on_eval,str) else None)

Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: fixed

File: fastestimator/estimator/trace.py [hunk 9]
```diff
@@ -1045,27 +1076,25 @@ class ModelCheckpoint(Trace):
...
         mode (str): Save models during either `train` or `eval`. Default is `eval`.
-        verbose (int): verbosity mode, 0 or 1
         save_best_only (bool): Keep only the best model.
         save_mode: Can be `'min'`, `'max'`, or `'auto'`.
         save_freq: Number of epochs to save models. Cannot be used with `save_best_only=True`.
     """
     def __init__(self,
                  save_dir,
-                 model_names=None,
+                 models,
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: is models really more intuitive than model_names here? if so, the doc string needs to say 'models' rather than 'model'

File: fastestimator/estimator/trace.py [hunk 10]
```diff
@@ -1045,27 +1076,25 @@ class ModelCheckpoint(Trace):
...
                  monitor_name=None,
                  mode='eval',
                  save_best_only=False,
                  save_mode='auto',
-                 save_freq=1,
-                 verbose=1):
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: verbose was probably here to match the keras method signature. how much to we care about keeping that the same? if we don't care we can probably simplify some other traces too

File: fastestimator/network/lrschedule.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        self.total_epochs = None  #will be filled by runtime
+        self.total_steps = None  #will be filled by runtime
+        self.initial_lr = None  #will be filled by runtime
+
+    def schedule_fn(self, current_step_or_epoch, lr):
+        #do something to change lr
+        return lr
+
+
+class CyclicLRSchedule(LRSchedule):
```
Comment (TortoiseHam) [inline fastestimator/network/lrschedule.py]: do you think it would be worth implementing the math functions here as @tf.function since they might get invoked at every step?

File: fastestimator/estimator/trace.py [hunk 12]
```diff
@@ -195,6 +195,97 @@ def on_epoch_end(self, state):
...
+                    self.change_lr = True
+                    print("FastEstimator-LRController: learning rate reduced by factor of {}".format(
+                        self.reduce_factor))
+
+    def _update_lr(self):
+        self.current_lr = max(self.base_lr * self.reduce_lr_ratio, self.min_lr)
+        backend.set_value(self.model.optimizer.lr, self.current_lr)
+        self.change_lr = False
+
+    def _is_better(self, value):
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: this could be done by selecting the comparison op up front when you're setting reduce_metric_best in the init method. see the early stopping trace for example

File: fastestimator/estimator/trace.py [hunk 13]
```diff
@@ -195,6 +195,97 @@ def on_epoch_end(self, state):
...
+    def on_batch_begin(self, state):
+        if state["mode"] == "train":
+            if self.lr_schedule and self.lr_schedule.schedule_mode == "step":
+                self.base_lr = self.lr_schedule.schedule_fn(state["train_step"], self.base_lr)
+                self.change_lr = True
+            if self.change_lr:
+                self._update_lr()
+
+    def on_batch_end(self, state):
+        if state["mode"] == "train" and self.log_steps and state["train_step"] % self.log_steps == 0:
```
Comment (TortoiseHam) [inline fastestimator/estimator/trace.py]: is there any reason to only emit this on log_steps? doing this if check might be just as expensive as just always putting it in the dictionary

File: apphub/image_classification/lenet_mnist.py (hunk 7 above)
Comment (TortoiseHam) [reply apphub/image_classification/lenet_mnist.py]: but you took it out of the trace list here. either it should be put back or the unused import should be removed

File: fastestimator/estimator/trace.py (hunk 13 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: yes, the performance improved when only emit this on log_steps

also, we don't want unnecessary stuff to be included in the state dictionary if they are not going to be be used anywhere.

File: fastestimator/estimator/estimator.py [hunk 14]
```diff
@@ -80,6 +82,11 @@ def _prepare_estimator(self):
             assert isinstance(trace, Trace)
             if isinstance(trace, ModelCheckpoint):
                 no_save_warning = False
+            if isinstance(trace, LRController):
+                trace.log_steps = self.log_steps
+                if trace.lr_schedule:
+                    trace.lr_schedule.total_epochs = self.epochs
```
Comment (TortoiseHam) [inline fastestimator/estimator/estimator.py]: instead of manually setting this from the estimator, we should pass these variables in the on_begin trace dictionary. that way other users who have traces that want this info can get it too

Comment (TortoiseHam) [reply fastestimator/estimator/estimator.py]: likewise with the log_steps info for all traces. better to minimize the amount of coupling between estimator and specific traces

File: fastestimator/estimator/estimator.py (hunk 1 above)
Comment (TortoiseHam) [reply fastestimator/estimator/estimator.py]: are we throwing out the last partial batch?

File: fastestimator/estimator/trace.py (hunk 13 above)
Comment (TortoiseHam) [reply fastestimator/estimator/trace.py]: huh. interesting

File: fastestimator/estimator/estimator.py (hunk 14 above)
Comment (vbvg2008) [reply fastestimator/estimator/estimator.py]: good idea, fixed

File: fastestimator/estimator/estimator.py (hunk 1 above)
Comment (vbvg2008) [reply fastestimator/estimator/estimator.py]: yes.

File: apphub/image_classification/lenet_mnist.py (hunk 7 above)
Comment (vbvg2008) [reply apphub/image_classification/lenet_mnist.py]: fixed

File: fastestimator/estimator/trace.py (hunk 9 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: fixed.

File: fastestimator/estimator/trace.py (hunk 10 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: yeah, we don't need to do the same, because we are already different from keras.

File: fastestimator/network/lrschedule.py (hunk 11 above)
Comment (vbvg2008) [reply fastestimator/network/lrschedule.py]: i don't think so, the computation is just a scalar, cpu is better for it.   

gpu is good for large data when data parallelism can be used.

File: fastestimator/estimator/trace.py (hunk 12 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: done

File: fastestimator/estimator/trace.py (hunk 13 above)
Comment (vbvg2008) [reply fastestimator/estimator/trace.py]: yup
